designs/disruption-controls-drain-timeout.md

```diff
@@ -0,0 +1,78 @@
+* Cluster admins who want to ensure that nodes are cycled after a given period of time, regardless of user-defined disruption controls (such as PDBs or preStop hooks) that might prevent eviction of a pod beyond the configured limit. This could be to satisfy security requirements or for convenience.
+* Cluster admins who want to allow users to protect long-running jobs from being interrupted by node disruption, up to a configured limit.
+
+This design piggybacks off the existing disruption-controls design, adding in a new `forceDrainAfter` field to the `v1beta/NodePool` `Disruption` spec. This field seems intuitively related to the other disruption controls that exist within the disruption block.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    forceDrainAfter: 24h || Never
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // ForceDrainAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable force draining of pods.
+    ForceDrainAfter *NillableDuration `json:"forceDrainAfter" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+The `forceDrainAfter` fields accepts a common duration string which defaults to a value of `Never`. Omitting the field results in the default value instructing the controller to wait indefinitely for pods to drain gracefully, maintaining the existing behavior.
+
+## Prior Art
+This is the field from CAPI's MachineDeployment spec which implements identical behavior.
+```
+$ kubectl explain machinedeployment.spec.template.spec.nodeDrainTimeout
+KIND:     MachineDeployment
+VERSION:  cluster.x-k8s.io/v1beta1
+
+FIELD:    nodeDrainTimeout <string>
+
+DESCRIPTION:
+     NodeDrainTimeout is the total amount of time that the controller will spend
+     on draining a node. The default value is 0, meaning that the node can be
+     drained without any time limitations. NOTE: NodeDrainTimeout is different
+     from `kubectl drain --timeout`
+```
+
+## Implementation
+
+### Force Drain Detection
+If forceDrainAfter is set, I think it should be straightforward to have the node termination controller add a timestamp as an annotation on the Node resource denoting when the Node will be force drained at a specific time (is RFC-3339 the preferred standard?)
+
+When a drain fails and would be requeued to be attempted again, if the current time is after the timestamp of the forceDrainAt annotation, a requeue should be skipped and the force drain logic should be triggered instead.
```

Comment:
I'm not sure I'm understanding this bit about requeueing. If a node has begun draining (i.e. begun terminating), then Karpenter treats this as past the point of no return. We'll keep trying indefinitely to evict the pods, and then once the timeout is reached, we'll delete the pods (not using the evictionAPI, thus not respecting PDBs), causing the pods to be terminated and the node eventually drained. 

One bit I think that we should include here is what the drain timeout means. Does it mean that after the timeout, if the node isn't fully drained, that we begin deleting all pods, respecting each of those pods grace period seconds? If not, should we proactively begin terminating pods in accordance with their grace period seconds so that they match up with the grace period timeout? I'm of the opinion that it should be the former, but curious to hear other opinions.

Reply:
> We'll keep trying indefinitely to evict the pods

My understanding is that this indefinite retry behavior is handled by a requeue mechanism when a drain fails, but I could be misunderstanding the controller's current behavior.

> Does it mean that after the timeout, if the node isn't fully drained, that we begin deleting all pods, respecting each of those pods grace period seconds?

No, a user can define an arbitrarily high `terminationGracePeriodSeconds` for their pods. In our use cases, users set the value to 24h and include a preStop hook that attempts to clean up and terminate long running jobs to minimize disruption during node rollouts. We use the node's `terminationGracePeriodSeconds` as a hard-stop mechanism that cannot be avoided. I think this behaves similarly to the pod's `terminationGracePeriodSeconds`, where you cannot do something like add a sleep greater than the grace period to prolong the life of the pod; it's a hard limit.

Reply:
So essentially this defines the contract for maximum length of drain to be `node.terminationGracePeriodSeconds + max(pod.terminationGracePeriodSeconds)`, is that right?

Reply:
Thats a good question. I think at this time the maximum length of a drain is just `node.terminationGracePeriodSeconds` and the assumption is within `node.terminationGracePeriodSeconds` all pods should be drained? (At least from what I understood)

Reply:
Yes, @garvinp-stripe has the correct understanding. A pod's `terminationGracePeriodSeconds` is not considered.

If a pod had a `terminationGracePeriodSeconds` of 2 hours, and the node's `terminationGracePeriod` was 1 hour, deletion of the node could cause a pod to be forcibly terminated an hour before the pod's `terminationGracePeriodSeconds` would expire.

Reply:
> If a pod had a terminationGracePeriodSeconds of 2 hours, and the node's terminationGracePeriod was 1 hour, deletion of the node could cause a pod to be forcibly terminated an hour before the pod's terminationGracePeriodSeconds would expire.

Does this consider the case where we have a pod that never receives an eviction request during the terminationGracePeriod for the node? For example: (1) pods with blocking PDBs, (2) do-not-disrupt pods, or (3) pods that are opted out of termination by default. In this 1 hour period, these pods would never receive a delete call, which means that as soon as forceful termination kicks in, these pods are _never_ given a chance to gracefully terminate. 

This is the core question I am still undecided about: what should the maximum time a node is terminating be? I can imagine taking a stance of being application-safe, by saying that it's `node.terminationGracePeriodSeconds + max(pod.terminationGracePeriodSeconds)`, but can also imagine that users who don't want to allow applications to extend the lifetime of a node to say it just means `node.terminationGracPeriodSeconds`, which may result in the pod terminationGracePeriodSeconds a best effort setting. in the case of spot instance reclamation, this is how it's treated.

Reply:
That's an interesting point. At my company we handle that case by preemptively identifying situations where Deployments and other resources are at their "PDB threshold" for extended periods of time, which is considered unusual behavior. You'd want to be checking for those situations as a workload maintainer in K8s regardless of this feature existing or not. You also have to tolerate the case where pods are terminated without a proper SIGTERM, because of power loss scenarios and other extenuating cases.

I think we can put some guidance in the docs for this feature to caution cluster admins that applications with startup times longer than the `terminationGracePeriod` could be at high risk for eviction without proper graceful pod termination every time nodes are cycled.

As the entire point of this feature is to be a hard limit that cannot be avoided from the user side, I think any guarantee based on a pod's `terminationGracePeriod` is a non-starter.

Reply:
To add to this point. I agree this is dangerous however there is a risk of letting users "set" a termination deadline and I think its a more problematic approach. We give users a very specific SLA on how long a node will provide a graceful termination period before we hard shut down the node. The reason for this is we have compliance standards to meet. For example, its pretty common for companies to enforce daily or weekly node refresh to minimize, in the case of a compromised user application, the time a compromised application has access to the node. It also allows us to enforce an SLA on when nodes will get updates. 

IMO, allowing cluster operators to manage this risk rather than not is a better approach. I think its unreasonable to let application users to dictate a host's graceful termination and instead cluster operator should set SLA for users on their platform.  We definitely should clearly state that this hard termination period ignores any pod level termination period for those who don't operate the same way.

Reply:
I can totally understand both of your perspectives here, where as cluster admins, you essentially want an override over application developers who may be acting in bad practice. The difference I think we should call out is that when Karpenter has a choice, it always aims to maximize application availability. 

For instance:
- Disruption for Expiration/Drift/Consolidation won't disrupt/replace a node if it isn't able to re-schedule _all_ of the pods on the node. 
- Karpenter will wait for a replacement node to initialize before deleting nodes so that pod downtime can be minimized. 

Generally, I agree that it would be nice to have a terminationGracePeriod that can be a sure-fire way to know "after x amount of time, I'll have this node cleaned up, regardless of what's on it" but I worry about the implications of this with scheduling. 

Users may need to be more rigorous about the nodepools that have scheduling constraints compatible with their pods but may not have terminationGracePeriods that are compatible with their applications graceful termination period. If they had a pod that was accidentally scheduled to a nodePool with a shorter terminationGracePeriod, the pod could still be running, but would not have enough time to cleanup. Ensuring that the pods always have a say here will ensure that applications clean up with the allotted time they're given, regardless of how they are scheduled. We've made this situation more likely with NodePool weights, where users can define a "preference" of nodePool schedulability, which may not map perfectly with terminationGracePeriod. 

Additionally, not respecting the terminationGracePeriod of pods that want to be the last to run on a node, which may not be included in the core drain function, will never get a chance to terminate, should we choose to implement it like this, which we've thought about implementing with a `NoExecute` taint. Adding a NoExecute taint at the end of the terminationGracePeriod may be a better k8s-native way to handle this logic too. 

I would be curious to know what @sftim 's thoughts are on this thread.

Reply:
The current implementation PR makes use of the official "out-of-service" NoExecute taint (requiring K8s 1.26 or feature gates). https://github.com/kubernetes-sigs/karpenter/pull/916
https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown

Reply:
> requiring K8s 1.26 or feature gates

IIUC, this should still work for all versions of K8s; the non-graceful feature just gives us the extra benefit of the node getting force-drained if the node's taint eviction manager is down. So, for the vast majority of cases, this should handle the same as a typical eviction flow when a `NoExecute` is added.